### PR TITLE
fix(gatsby-transformer-sharp): removed unnecessary conversion to webp

### LINKED ISSUE #14497: [gatsby-transformer-sharp] Transparent PNG image transform to Webp does not create correct image
## Description

Transparent PNG image transform to Webp does not create correct image.

I have tested with the following script

```js
const sharp = require("sharp");

sharp("docker.png").webp().toFile("output.webp");
```

This produce a correct Webp image which make me believe this is a Gatsby problem.

### Steps to reproduce

```
{
  resolve: "gatsby-remark-images",
  options: {
    maxWidth: 1200,
    withWebp: true,
    quality: 85,
    tracedSVG: false,
    showCaptions: true
  }
}
```

### Environment

```
  System:
    OS: Windows 10
    CPU: (8) x64 Intel(R) Core(TM) i7-4790K CPU @ 4.00GHz
  Binaries:
    npm: 6.9.0
  Languages:
    Python: 3.7.0
  Browsers:
    Edge: 42.17134.1.0
  npmPackages:
    gatsby: ^2.7.5 => 2.7.5
    gatsby-image: ^2.1.2 => 2.1.2
    gatsby-plugin-manifest: ^2.1.1 => 2.1.1
    gatsby-plugin-offline: ^2.1.1 => 2.1.1
    gatsby-plugin-react-helmet: ^3.0.12 => 3.0.12
    gatsby-plugin-sass: ^2.0.11 => 2.0.11
    gatsby-plugin-sharp: ^2.1.2 => 2.1.2
    gatsby-plugin-typescript: ^2.0.15 => 2.0.15
    gatsby-remark-external-links: 0.0.4 => 0.0.4
    gatsby-remark-images: ^3.0.14 => 3.0.14
    gatsby-remark-prismjs: ^3.2.9 => 3.2.9
    gatsby-source-filesystem: ^2.0.37 => 2.0.37
    gatsby-transformer-remark: ^2.3.12 => 2.3.12
    gatsby-transformer-sharp: ^2.1.20 => 2.1.20
```

### PR DESCRIPTION
**Pull request by ungvert**
## Description

In process-file.js conversion to `webp` happends two times:
- With `Sharp` after `.resize()` and `.png()`
- With `imageminWebp` in `compressWebP()`

If we convert `png` image with transparency to `webp` and then we pass this `webp` image to `imageminWebp` - it returns image with weird artefacts and black background (tested on Windows 10).  

I don't comprehend why we doing the conversion two times. Parameters of conversion seems identical.

## Related Issues
Fixes #14497

### CODE REVIEW

File: packages/gatsby-plugin-sharp/src/process-file.js [hunk 1]
```diff
@@ -116,10 +116,6 @@ exports.processFile = (file, transforms, options = {}) => {
           adaptiveFiltering: false,
           force: transformArgs.toFormat === `png`,
         })
-        .webp({
-          quality: transformArgs.webpQuality || transformArgs.quality,
-          force: transformArgs.toFormat === `webp`,
-        })
```
Comment (polarathene) [inline packages/gatsby-plugin-sharp/src/process-file.js:122]: I think this should be the other way around?

I've recently noticed that my output from processing with `sharp` without gatsby differed in quality and filesize for some reason. Came across your PR here, and now it makes sense that it's due to the double compression as you've found here, but imagemin is the culprit.

It reduces filesize from 16.5KiB to 15.6KiB, but introduces some notable artifacts, both would be running at quality of 75. When removing the `sharp` processing that you've done here, just with imagemin instead, the output filesize is 17.1KiB for my test file and the quality is not as good.

Comment (ungvert) [reply packages/gatsby-plugin-sharp/src/process-file.js]: Yes, you right. It should be other way around - I should have been deleted second conversion with imageminWebp. I fixed it in new commit.

File: packages/gatsby-plugin-sharp/src/process-file.js [hunk 2]
```diff
@@ -229,8 +229,11 @@ const compressJpg = (pipeline, outputPath, options) =>
...
 const compressWebP = (pipeline, outputPath, options) =>
-  pipeline.toBuffer().then(sharpBuffer =>
-    imagemin
+  pipeline.toBuffer().then(sharpBuffer => {
+    if (process.platform === `win32`) {
+      return fs.writeFile(outputPath, sharpBuffer)
+    }
+    return imagemin
       .buffer(sharpBuffer, {
         plugins: [
```
Comment (polarathene) [inline packages/gatsby-plugin-sharp/src/process-file.js:238]: What's going on here? Why have you modified this function to write the buffer to disk on Windows but continue the problem on other platforms?

Don't modify this method, it does not need to be run, the file will still be written to disk. Instead the condition earlier that calls the method should only call it if the user has configured it to do so(either globally in `gatsby-config` or via a new graphql param `compress`/`minify` and corresponding format only variants like `quality` does. Assuming @wardpeet agrees with that approach first.

**Comment by ungvert** [comment]
@polarathene I'm sorry. I didn’t have time to respond fully to you yesterday.

I made a commit that solves the base case of the issue. I myself encountered this problem with artifacts on my Gatsby project, when i tried to convert a png file with transparency to webp on Windows. In issue, no one talked about other operating systems at all, on the contrary, people mentioned that when deploying to Linux(Netlify) images was fine.

I understand that solving this problem, new questions arise:

- Why the conversion occurs two times. By the way, not only webp, png is also converted 2 times. And with different libraries. 
- Do we need to leave a second conversion as is? Do we have a calculated benefits of second conversion?
- If second conversion would be optional, how should we implement it?


Should all this be resolved within this pull request?

I see several tasks here, and I think that they probably should be solved in separate issues.

If the core team decide that it should be part of this pull request -  then ok, we can discuss that.

But for now my fix solves the base problem that was araised in the issue. And i think it would be beneficial for Windows users to have that.

**Comment by polarathene** [comment]
> But for now my fix solves the base problem that was raised in the issue. And i think it would be beneficial for Windows users to have that.

I understand that, but I'm not fond of the solution you're provided for the problem(Transparent PNG to Webp breaking on Windows), as it doesn't address inconsistency across OS, it just changes the output on Windows(which wouldn't be affected under any other Webp output scenario).

> In issue, no one talked about other operating systems at all, on the contrary, people mentioned that when deploying to Linux(Netlify) images was fine.

The issue I've raised here differs, unrelated to the transparency bug; encoding an image twice in a row has compounding effect on compression quality, where in my test case had minimal file size reduction but notable artifacting(blocky/pixelated visuals) which would be avoided by just lowering quality further in a single pass instead of the current double pass. 

This was experienced on Linux but would also be valid on other OS too, visibility of the artifacting likely depends on image data(notable one for me was low frequency details of gradient/blurred colours).

---

> Why the conversion occurs two times. By the way, not only webp, png is also converted 2 times. And with different libraries.

All images are being processed via Sharp first which handles other image operations like cropping and duotone. The 2nd pass through imagemin is purely meant to optimize the image further, akin to css/js/html minifiers. If the quality was kept at 100 for that, the image data shouldn't degrade further presumably, but this also seems to potentially raise file size, defeating it's purpose as a "compression" pass.

> Do we need to leave a second conversion as is? Do we have a calculated benefits of second conversion?

We would need a large enough image corpus to measure benefits at a variety of settings/sizes. I imagine it was done with the intention of a best practice, and possibly evaluating file size reduction without taking an image diff or structural similarity(SSIM), [this article](https://compress-or-die.com/The-big-image-compression-tool-comparison) might give some insights on how to approach it, it's just a question of who has the time to actually go through that.. I'm doubtful anyone cares enough to do so, at least on a voluntary basis.

> If second conversion would be optional, how should we implement it?

Opt-in via graphql param or global sharp setting as I mentioned earlier.

eg with `compression: true` in graphql:

```graphql
childImageSharp {
    fluid(maxWidth: 512, compression: true) {
      ...GatsbyImageSharpFluid_withWebp
    }
}
```

and `gatsby-config.js`:

```js
{
  resolve: "gatsby-plugin-sharp",
  options: {
    compression: true,
  },
},
```

One can look up my base64 PR for custom size to see how to implement that. I'm happy to make the PR. 

In the `process-file.js`, you would access the `compression` option just like `toFormat` is accessed. It can be used in the conditionals that call imagemin compression functions to only do so if compression is enabled.

---

If you like, I can create that PR and we can close this one. Otherwise, you might as well alter this PR to strip out compression passes, leave that to be implemented by a separate plugin or optimization deferred to the user to run over their build output.

**Comment by polarathene** [comment]
I've gone through the history via `git blame` and tracked down [where imagemin was added](https://github.com/gatsbyjs/gatsby/blob/e51dc4fab63e5478758d90634ba56875beb51a32/packages/gatsby-sharp/src/index.js#L8), the commit message and [large PR](https://github.com/gatsbyjs/gatsby/pull/662/commits/e51dc4fab63e5478758d90634ba56875beb51a32) isn't particularly insightful. This change was done at the end of 2016, viewing the PR individual commit history, a TODO message is found for **why imagemin was added**:

https://github.com/gatsbyjs/gatsby/blob/1f418535f093926e617d2826183fcd40e1c634d2/packages/gatsby-sharp/index.js#L36-L37

That's the sole purpose we have it.

In 2018, that [still appeared to be the case](https://github.com/lovell/sharp/issues/1183) with a lack of compression support for PNG. Looking at the [changelog for sharp](https://sharp.pixelplumbing.com/changelog), we can see support for libimagequant arrived at Jan 2019, along with metadata stripping(default) and in June 2020 a further optimization for indexed palettes([auto enabled](https://github.com/lovell/sharp/pull/2226) if quality/colors/dither options are set). 

`libimagequant` needs to be provided on the system, it's not distributed with libvips or sharp binaries due to GPL license(there is a way to [get a linux binary via env var however](https://github.com/lovell/sharp/issues/1484#issuecomment-494145836)). That is required to benefit from optimizations like the indexed palettes, as well as [the `quality` setting having any actual impact apparently for PNG via sharp](https://github.com/lovell/sharp/issues/1793#issuecomment-572983865).

Thus imagemin may still be relevant for compressing PNGs.

---

[mozJPEG as an alternative encoder](https://github.com/gatsbyjs/gatsby/pull/8621/files) for providing better compression. This is supported in a manner that it only uses sharp or mozJPEG, not both.

When [imagemin webp was added](https://github.com/gatsbyjs/gatsby/pull/2560), presumably under the assumption that since other formats used imagemin, this should too. Kyle completely overlooked it himself while reviewing the PR.. There doesn't [appear to be any reason/benefit to use this plugin](https://github.com/imagemin/imagemin-webp).

The imagemin provides a variety of alternative PNG options as well as some for jpeg, the addition of imagemin-webp was likely a a mistake so it should be removed entirely.

**Comment by ungvert** [comment]
@polarathene Wow. Great investigation! I understand more clearer what you meant in previous comments. Thank you for that!

I made new commit which deletes everything related to `imagemin-webp`.

**Pull request merged**